avocado/utils/pmem.py

```diff
@@ -0,0 +1,298 @@
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright (C) IBM 2020 - Harish <harish@linux.vnet.ibm.com>
+#
+# Author: Harish <harish@linux.vnet.ibm.com>
+#
+
+
+import re
+import json
+import glob
+import logging
+
+from . import process
+from . import genio
+
+log = logging.getLogger('avocado.test')
+
+
+class PMemException(Exception):
+    """
+    Error raised for all PMem failures
+    """
+    def __init__(self, additional_text=None):
+        self.additional_text = additional_text
+
+    def __str__(self):
+        return ("Command failed.\ninfo: %s" % self.additional_text)
+
+
+class PMem:
+    """
+    PMem class which provides function to perform ndctl and daxctl operations
+
+    This class can be used only if ndctl binaries are provided before hand
+    """
+
+    def __init__(self, ndctl='ndctl', daxctl='daxctl'):
+        """
+        Initialize PMem object
+
+        :param ndctl: path to ndctl binary, defaults to ndctl
+        :param daxctl: path to daxctl binary, defaults to ndctl
+        """
+        self.ndctl = ndctl
+        self.daxctl = daxctl
+
+    def run_ndctl_list(self, option=''):
+        """
+        Get the json of each provided options
+
+        :param option: optional arguments to ndctl list command
+        :return: By default returns entire list of json objects
+        :rtype: list of json objects
+        """
+        try:
+            json_op = json.loads(process.system_output(
+                '%s list %s' % (self.ndctl, option), shell=True))
+        except ValueError:
+            json_op = []
+        return json_op
+
+    @staticmethod
+    def run_ndctl_list_val(json_op, field):
+        """
+        Get the value of a field in given json
+
+        :param json_op: Input Json object
+        :param field: Field to find the value from json_op object
+
+        :rtype: Found value type, None ig not found
+        """
+        for key, value in json_op.items():
+            if key == field:
+                return value
+        return None
+
+    def run_daxctl_list(self, options=''):
+        """
+        Get the json of each provided options
+
+        :param options: optional arguments to daxctl list command
+        :return: By default returns entire list of json objects
+        :rtype: list of json objects
+        """
+        return json.loads(process.system_output(
+            '%s list %s' % (self.daxctl, options), shell=True))
+
+    def get_slot_count(self, region):
+        """
+        Get max slot count in the index area for a  dimm backing a region
+        We use region0 - > nmem0
+
+        :param region: Region for which slot count is found
+        :return: Number of slots for given region
+        :rtype: int
+        """
+        nmem = "nmem%s" % re.findall(r'\d+', region)[0]
```

Comment:
what about if  region is not valid library should break can we have that check as well as in this case the library will fail I see try for actual cmd run my point is if re itself give a bogus value, seems me we need to handle that

Reply:
region being an input to this function, is upto the caller to provide right value. When you say region is not valid you mean wrong input ? The caller always is expected to provide a proper input as the result of the function will subsequently used.

Reply:
yes it wrong region input, yes it should be always a valid one but in wrong input, the library needs to be handled

Reply:
ok.